### ISSUE TITLE
[PiranhaJava] Does not remove Java enum constants with fields matching flag value

### ISSUE THREAD

**Issue by tnterdan** [uber/piranha#141]
So let's say I have an enum with some code like this:

```java
public enum TestExperimentName {
    STALE_FLAG("stale.flag"),
    OTHER_FLAG("other.flag");
    private final String key;
    TestExperimentName(final String key) {
        this.key = key;
    }
}
```

Running Piranha with `-XepOpt:Piranha:FlagName=stale.flag` won't remove the `STALE_FLAG` constant from the enum (and therefore will also not remove downstream usages).

I have been working on a fix for this, as my organization uses feature flags in this way. Still waiting on legal approval before I can share any code with y'all, however. And AFAIK, it will require its own pass and my current implementation has its own BugChecker implementation.

Would love to know if the maintainers of the Piranha project believe this is out of scope or not for Piranha. While I'd rather contribute to this back upstream to the project, I can see it being arguably made to be in its own project to be run prior to running Piranha, as it is a use case quite possibly specific to my organization's needs.

**Comment by mkr-plse** [comment]
This is in scope for Piranha. It will be ideal to refactor our implementation to support various styles of flag definitions by decoupling the code cleanup from flag definition inference. 

Can you provide a bit more context on how `key` can be associated with multiple flags (`stale.flag`, `other.flag`)? Some examples of the use of `key` with `STALE_FLAG` and `OTHER_FLAG` will be helpful.

**Comment by tnterdan** [comment]
Thanks for the quick reply @mkr-plse. Glad to hear it's in scope, will be happy to share code once legal approves then.

Let me provide some context. We use LaunchDarkly for our feature flags. We define them in LaunchDarkly using strings in the format `product-name.flag.name`, which doesn't map cleanly to enum constant names. So we had to come up with using the enum `key` field to associate the flag usage in the codebase with the flag usage in LaunchDarkly if we wanted to use enums for flags.

The `key` is most of the time to only be associated with the enum in our `FeatureFlag.java` file, and then used throughout the codebase using its corresponding enum. So for example we have our `FeatureFlag.java`, pretty much as above, but with hundreds of flags instead of 2:
```java
public enum FeatureFlag {
    // ... a hundred+ other flags
    STALE_FLAG("stale.flag"), // omit product-name prefix, as it is the same for all flags
    OTHER_FLAG("other.flag");

    private final String key;

    public String getKey() {
        return key;
    }

    FeatureFlags(final String key) {
        this.key = key;
    }
}
```
`key` should be unique inside `FeatureFlag` in our case, but there's no reason I can think of that Piranha couldn't handle a case where there were two enums constants that had identical key values.

Then we might have a test that uses that flag, using a feature flag client class to handle evaluation. As you can see, we pass the key String rather than the enum itself:
```java
public void testExample() {
    if (featureFlagClient.isFeatureEnabled(FeatureFlag.STALE_FLAG.getKey(), /* default = */ false)) {
        // test if flag is enabled...
    } else {
        // test if flag is disabled...
    }
}
```

Although there are times where we use both the enum and directly using the key String for other usages. So other parts of the code might use the name of the flag directly, likely written before we had `FeatureFlag.java` as a place to put all flags:
```java
public void doSomething() {
    if (featureFlagClient.isFeatureEnabled("stale.flag", /* default = */ false)) {
        // do something if flag is enabled...
    } else {
        // do something if flag is disabled...
    }
}
```

**Comment by lazaroclapp** [comment]
So, we support enum flags and string flags, and then allow a series of methods and annotations to be specified inside `properties.json` so that Piranha takes actions on them if they reference the flag. These actions are themselves defined by the configuration inside `properties.json`.

My thoughts here are that this might just be a case of "string valued flags", and the feature that we need to add is a `properties.json` configuration for enums which specifies: a) an unambiguous way to find out which string flag is associated with the enum (for enum deletion), b) a list of the methods used to retrieve the string value from the enum (for simplification of flag usages in your `FeatureFlag.STALE_FLAG.getKey()` example above).

This should be doable without an extra Error Prone pass, I think. But I am not 100% sure about that. Also, it should be done in a way that doesn't change existing Piranha behavior when enum specifications are not explicitly defined in the configuration. That said, we are definitely interested and happy to review a PR around this issue and discuss it from there.

**Comment by mkr-plse** [comment]
+1. Let us add it as part of the same EP pass (if necessary, under an option).

**Comment by tnterdan** [comment]
Hmm... I suppose if we can do a look up for every enum's String field values or constructor when Piranha encounters that enum being used, it would be possible to do in one pass. Let me see what kind of tools I can use from Error Prone to traverse the AST and I'll get back to y'all with a PR with a new option in the `properties.json` for this. And it looks like legal just approved me for code sharing :)

Really appreciate the replies @lazaroclapp and @mkr-plse !
